Escaped characters detection in literals

ferranpujolcamins opened the pull request:
Followup of https://github.com/pointfreeco/swift-snapshot-testing/pull/228, Fixes https://github.com/pointfreeco/swift-snapshot-testing/issues/230

This PR fixes inline snapshot tests containing special characters. AssertInlineSnapshot now checks whether the snapshot contains special characters or not. If it does, escapes the snapshot literal appropriately. 

Known issue: AssertInlineSnapshot still searches only for "" and multiline string literals (it does not search #""" """# for example)

stephencelis commented on SnapshotTesting.xcodeproj/project.pbxproj:
```diff
@@ -647,7 +671,11 @@
 				BUNDLE_LOADER = "$(TEST_HOST)";
 				INFOPLIST_FILE = Tests/Info.plist;
-				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks @loader_path/Frameworks";
+				LD_RUNPATH_SEARCH_PATHS = (
```
We auto-create the Xcode project file using XcodeGen. Can you run `make xcodegen` from the project directory and recommit that file?

ferranpujolcamins replied:
Done!

stephencelis commented on Sources/SnapshotTesting/Common/String+SpecialCharacters.swift:
```diff
@@ -0,0 +1,61 @@
+//
```
Cool to remove this header comment? We delete them in all of the other files.

ferranpujolcamins replied:
Sure!

stephencelis commented on Tests/SnapshotTestingTests/String+SpecialCharactersTests.swift:
```diff
@@ -0,0 +1,133 @@
+//
```
Would you please remove this header comment, too?

ferranpujolcamins commented:
> Do you think you'll have time to address this soon?

I'm giving it a try

ferranpujolcamins commented:
> One interesting side effect is that it'll automatically insert more #s than are needed when re-recording

I managed to make it work. Not without pain! Substrings are tricky. Mutating self in `AssertInlineSnapshot.replaceFirstOccurrence` produced crazy results, I don't exactly know why.

stephencelis commented:
Hey @ferranpujolcamins! Looks like we're close but not _quite_ there yet. If you run the following test block:

``` swift
  func testAny() {
    struct User { let id: Int, name: String, bio: String }
    let user = User(id: 1, name: ##"Bl#""#obby"##, bio: "Blobbed around the world.")
    assertSnapshot(matching: user, as: .dump)
    _assertInlineSnapshot(matching: user, as: .dump, with: #####"""
    """#####)
  }
```

The leading `#####"""` correctly becomes `##"""`, but the trailing `"""#####` ends up appending an extra coupla `#`s to become `"""#######`. I think the problem is the use of `multiLineLiteralEndIndex` for replacement, which only matches against `"""`.

The logic you're building on is definitely the most complicated part of the lib, sorry! Let us know if we can help in any way.

ferranpujolcamins commented:
I see! This won't be an easy change after all. Let's do this:
I'll open other PRs with small refactors that will make our live easier. My goal is to be able to unit test the inline snapshot feature, so I will separate code transformation from file operartions.

After that, we can keep working on this one.

ferranpujolcamins commented:
Would be cool to add swift-prelude as dependency?

ferranpujolcamins commented:
Mmmm I see it depends on this package so we can't use it here :(

ferranpujolcamins commented:
Ok, I've rebased onto master.
I've added some tests. Your counter example is now working properly.
I still need to add some more, but I'd appreciate early feedback.

The tests work as follows (it's a little bit confusing):
The tests in `InlineSnapshotTests` assert the correctness of the source file modifications that the inline snapshot test machinery does. To do so, we snapshot test the modified source.

The recorded snapshots, are in turn well formed tests that are added to the test target.
* Thus, we check the source code after an inline snapshot is recorded is valid. 
* They assert that the recorded snapshot is escaped properly.
Please see comment at the end of `InlineSnapshotTests.swift`.

stephencelis commented on Tests/SnapshotTestingTests/InlineSnapshotTests.swift:
```diff
@@ -4,70 +4,411 @@ import XCTest
+    assertSnapshot(source: newSource, diffable: diffable)
+  }
+
+  // TODO: add more tests like this with different amount of lines
```
Just wanna flag this before merging.

stephencelis commented on Tests/SnapshotTestingTests/InlineSnapshotTests.swift:
```diff
@@ -4,70 +4,411 @@ import XCTest
+  }
+}
+
+// Class that is extended with the generated code to check that it builds.
```
Cool stuff!

ferranpujolcamins commented:
Yeah I just added a couple of tests. Should be ready to merge. If we find some corner case later let's add a regression test!

Thanks guys!

ferranpujolcamins commented:
Gentle ping ;)

stephencelis commented:
Beautiful work! This'll eliminate a major gotcha for folks. Thanks again for contributing!

ferranpujolcamins commented:
Thank you guys for working on this useful project!
